### ISSUE TITLE
RecursionError during call to T2CharString.decompile()

### ISSUE THREAD

**Issue by cjchapman** [fonttools/fonttools#1560]
In the attached test font `subr_test_font_infinite_recursion.otf` the T2CharString for glyph `F` calls a global subroutine that in turn calls itself.

When I try to decompile the T2CharString for this glyph using the `decompile()` method, I get:

```
Traceback (most recent call last):
  File "/Users/cchapman/bin/decompile_cff_charstrings.py", line 13, in <module>
    t2_char_string.decompile()
  File "/usr/local/lib/python3.7/site-packages/fontTools/misc/psCharStrings.py", line 966, in decompile
    decompiler.execute(self)
  File "/usr/local/lib/python3.7/site-packages/fontTools/misc/psCharStrings.py", line 300, in execute
    rv = handler(index)
  File "/usr/local/lib/python3.7/site-packages/fontTools/misc/psCharStrings.py", line 343, in op_callgsubr
    self.execute(subr)

  ... I've removed a couple thousand more lines of the same for the sake of brevity ...

  File "/usr/local/lib/python3.7/site-packages/fontTools/misc/psCharStrings.py", line 300, in execute
    rv = handler(index)
  File "/usr/local/lib/python3.7/site-packages/fontTools/misc/psCharStrings.py", line 343, in op_callgsubr
    self.execute(subr)
  File "/usr/local/lib/python3.7/site-packages/fontTools/misc/psCharStrings.py", line 292, in execute
    token, isOperator, index = charString.getToken(index)
  File "/usr/local/lib/python3.7/site-packages/fontTools/misc/psCharStrings.py", line 1040, in getToken
    if index >= len(self.bytecode):
RecursionError: maximum recursion depth exceeded while calling a Python object
```

Here's some example code to reproduce the problem:

```python
import sys
from fontTools.ttLib import TTFont

font_path = sys.argv[1]
tt_font = TTFont(font_path)
cff = tt_font['CFF '].cff
for top_dict in cff.topDictIndex:
    char_strings = top_dict.CharStrings
    char_list = char_strings.keys()
    for key in char_list:
        print('glyph: %s' % key)
        t2_char_string = char_strings[key]
        t2_char_string.decompile()
        print(t2_char_string.program)
```

Just to be clear: yes, this is a deliberately bad test font with infinite recursion, but I'm reporting this because I was expecting `decompile()` to simply decompile the charstring. I wasn't expecting it to execute the subroutines called by that charstring.

[subr_test_font_infinite_recursion.otf.zip](https://github.com/fonttools/fonttools/files/3014560/subr_test_font_infinite_recursion.otf.zip)

**Comment by cjchapman** [comment]
This is a `CFF2` font I created with the same idea of glyphs `B` - `E` having call depths 9-12, and glyph `F` calling a `gsubr` that calls itself:
[var_subr_test_font_infinite_recursion.otf.zip](https://github.com/fonttools/fonttools/files/3014904/var_subr_test_font_infinite_recursion.otf.zip)

...it also results in `RecursionError` when trying to decompile the charstring for the `F`.

**Comment by justvanrossum** [comment]
Due to how `hintmask` works it's not reliably possible to decompile without going into the subroutines, so I'm not sure I see a clean solution to this.

On the other hand, perhaps we can improve the error message, like we did for #1545 in #1546.

**Comment by cjchapman** [comment]
@justvanrossum Oh, right, that makes sense. I'm sorry that I didn't think about the need for the stem hints counts from any subroutines called to know how many data bytes follow a `hintmask` when I wrote this up.

I was working on a call depth checker for Font Bakery when I ran into this, and I just did this in my code:
```python
            try:
                t2_char_string.decompile()
            except RecursionError:
                yield FAIL, "Recursion error while " \
                            "decompiling glyph '{}'.".format(key)
                failed = True
                continue
```

...and was enough to meet my needs.

I personally don't need the error message to be any more detailed, and I now understand why the decompile method is going into the subroutines, so if you want to close this issue, that would be fine with me.
